[Feature] [Balance] Add critical captures, update shake probability to match gen 6

Fontbane opened the pull request:
<!-- Make sure the title includes categorization (i.e. [Bug], [QoL], [Localization]) -->
<!-- Make sure that this PR is not overlapping with someone else's work -->
<!-- Please try to keep the PR self-contained (and small) -->

## What are the changes the user will see?
<!-- Summarize what are the changes from a user perspective on the application -->
- Players will now have a chance to get a [critical capture](https://bulbapedia.bulbagarden.net/wiki/Catch_rate#Critical_capture)
  - Uses the same formula as in the main games
  - Depends on number of unique species caught, and the calculated catch rate
    - The multiplier based on number of species caught is as follows:
      | Species Caught | Multiplier |
      | --- | --- |
      | >800 | 2.5 |
      | 601-800 | 2 |
      | 401-600 | 1.5 |
      | 201-400 | 1 |
      | 101-200 | 0.5 |
      | ≤100 | 0 |
  - Daily Runs use the max multiplier of 2.5x
  - Ball will shake in the air and drop to the ground, only performing one shake check (always shakes once)
  - Critical captures are **disabled in Fresh Start challenge**.
- Also updates the shake probability to match gen 6
  - Shake probability is what determines if the ball shakes each of the 3 times before catching
  - Shake checks occur 4 times, one for each shake and then one to see if the catch is successful
  - Critical captures bypass the first shake check and will always perform one shake, then do a shake check to see if the catch is successful.

## Why am I making these changes?
<!-- Explain why you decided to introduce these changes -->
<!-- Does it come from an issue or another PR? Please link it -->
<!-- Explain why you believe this can enhance user experience -->
- Make catching easier or at least up to date with current mechanics
- Catching is much harder in rogue due to abundance of berries and lack of specialized balls
- Critical captures are a neat feature and reward for progression
- Lays the groundwork for adding the Catching Charm

## What are the changes from a developer perspective?
<!-- Explicitly state what are the changes introduced by the PR -->
<!-- You can make use of a comparison between what was the state before and after your PR changes -->
- Adds a new function in `pokeball.ts`, `getCriticalCaptureChance`
- Adjusts `AttemptCapturePhase` to include the crit check and update shake probability chance
- Adds animation for critical capture in `doPokeballBounceAnim`

### Screenshots/Videos
<!-- If your change is changing anything on the user experience, please provide visual proofs of it -->
<!-- Please take screenshots/videos before and after your changes, to show what is brought by this PR -->
https://github.com/user-attachments/assets/43e8474f-811e-4f3c-859d-11452d801ff6

## How to test the changes?
<!-- How can a reviewer test your changes once they check out on your branch? -->
<!-- Did you just make use of the `src/overrides.ts` file? -->
<!-- Did you introduce any automated tests? -->
<!-- Do the reviewer need to do something special in order to test your change? -->

## Checklist
- [X] **I'm using `beta` as my base branch**
- [X] There is no overlap with another PR?
- [X] The PR is self-contained and cannot be split into smaller PRs?
- [X] Have I provided a clear explanation of the changes?
- [X] Have I considered writing automated tests for the issue?
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [X] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [X] Are the changes visual?
  - [X] Have I provided screenshots/videos of the changes?

PigeonBar commented:
Also, what is the intended behaviour of the critical capture check during Daily Runs, to avoid Daily Run inconsistencies between different players?

Tempo-anon commented:
Also fresh start since there's currently no way to remember run specific metaprogression within the context of a run

Fontbane commented:
Addressed all points.

- Disabled during Fresh Start and Daily Runs
- Critical captures skip the first shake check and are guaranteed to perform one shake. Then a shake check is done to see if it's caught.

DayKev commented on src/phases/attempt-capture-phase.ts:
```diff
@@ -80,30 +84,33 @@ export class AttemptCapturePhase extends PokemonPhase {
                 from: 0,
                 to: 1,
-                repeat: 4,
+                repeat: isCritical ? 2 : 4, // Critical captures only perform 2 shake checks
```
The page you linked says crit captures only do one shake check though?
> Once Y has been determined, the Pokémon will make up to four attempts to break out of the ball, except in a critical capture, where it will only make one.
- https://www.dragonflycave.com/mechanics/gen-vi-vii-capturing

Fontbane replied:
My bad forgot to change the comment

PigeonBar commented on src/phases/attempt-capture-phase.ts:
```diff
@@ -114,13 +121,18 @@ export class AttemptCapturePhase extends PokemonPhase {
                       shakeCounter.stop();
                       this.failCatch(shakeCount);
                     }
+                  } else if (isCritical && pokemon.randSeedInt(65536) >= shakeProbability) {
+                    // Above, perform the one shake check for critical captures after the ball shakes once
```
```suggestion
                  } else if (pokeballMultiplier !== -1 && pokemon.randSeedInt(65536) >= shakeProbability) {
                    // Perform shake check after final ball shake (for both critical and non-critical captures)
```
This will be the 4th shake check for non-critical captures

Fontbane replied:
Just realized, yes, we don't do a 4th shake check. I don't think we should though, catching is hard enough as is (this PR was meant to alleviate that) and I imagine it would generate bug reports if mons started breaking out after 3 shakes.

innerthunder commented on src/data/pokeball.ts:
```diff
@@ -82,11 +83,38 @@ export function getPokeballTintColor(type: PokeballType): number {
   let bounceY = y2;
   const yd = y2 - y1;
+  const x0 = pokeball.x;
+  const x1 = x0 + 4;
+  const x2 = x0 - 4;
```
```suggestion
  const x1 = x0 + 3;
  const x2 = x0 - 3;
```
Makes the critical capture shake less intense

innerthunder commented on src/data/pokeball.ts:
```diff
@@ -117,5 +145,40 @@ export function doPokeballBounceAnim(scene: BattleScene, pokeball: Phaser.GameOb
+                x: x0,
+                duration: 60,
+                ease: "Linear",
+                onComplete: () => doBounce()
```
```suggestion
                onComplete: () => scene.time.delayedCall(500, doBounce)
```
A slight delay to make the critical capture animation a bit more pronounced

innerthunder reviewed (commented):
Some animation fixes, as discussed in the Discord.

Video of the animation after the changes:

https://github.com/user-attachments/assets/0dcd63be-9a1a-4fc4-bda0-2aebcf4ca826
